contributors/design-proposals/scheduling/scheduling-policy.md

```diff
@@ -0,0 +1,302 @@
+  if policy matches pod: // all specified policy rules match
+    return policy.action
+```
+ note that rules of policies from a lower priority are superseeded by ones from a higher priority if they match.
+
+
+
+# Detailed Design
+
+
+## SchedulingPolicy
+
+Proposed API group: `policy/v1alpha1`
+
+
+### SchedulingPolicy content
+
+SchedulingPolicy spec is composed of optional fields that allow scheduling rules. If a field is absent from a SchedulingPolicy, this `SchedulingPolicy` won't allow any item from the missing field.
+
+```yaml
+apiVersion: policy/v1alpha1
+kind: SchedulingPolicy
+metadata:
+  name: mySchedulingPolicy
+spec:
+  priority: <exception,cluster,default>
+  action:  <allow,deny>     
+  rules:
+    namespaceSelector:
+      - key: value
+        operator: operator
+    podSelector:
+        key: value
+    schedulerNames:
+       match: []
+    priorityClassNames:
+       match: []
+    tolerations:
+       match: []
+    nodeSelectors:
+       match: []
+    nodeAffinities:
+       match: []
+    podAntiAffinities:
+       match: []
+    podAffinities:
+       match: []
+```
+
+- allow: When a policy matches the pod is allowed
+- deny: when a policy matches the pod is denied
+
+
+### Scoping
+
+The `spec.rules.namespaceSelector` and `spec.rules.podSelector` attributes scope the policy application range.
+
+Pods must match both `spec.namespaceSelector` and `spec.podSelector` matching rules to be constrained by a policy.
+Both `spec.namespaceSelector` and `spec.podSelector` are optional. If absent, all pods and all namespaces are targeted by the policy.
+
+### Scheduler name
+
+If `schedulerNames` is absent from `allowed`, `default` or `required`, no scheduler is allowed by this specific policy.
+
+
+#### Allowed
+Allow pods to use either the `green-scheduler` (which is used by specifying `spec.schedulerName` in pod definition) or the `my-scheduler` scheduler (by specifying `spec.schedulerName: "my-scheduler"`) in namespaces labeled `team`:
+
+```yaml
+apiVersion: policy/v1alpha1
+kind: SchedulingPolicy
+metadata:
+  name: mySchedulingPolicy
+spec:
+  action: allow    
+  rules:
+    namespaces:
+      - key: team
+        operator: Exists
+    podSelector: {}
+    schedulerNames:
+       match: ["green-scheduler","my-scheduler"]
+```
+
+An empty list of schedulerNames will allow usage of all schedulers:
+
+```yaml
+apiVersion: policy/v1alpha1
+kind: SchedulingPolicy
+metadata:
+  name: mySchedulingPolicy
+spec:
+  action: allow
+  rules:
+    namespaces:
+      - key: team
+        operator: Exists
+    podSelector: {}    
+    schedulerNames:
+       match: []
+```
+
+note: this policy has no effect as we allow all when no policy is set.
+
+### Tolerations
+
+Toleration usage can be regulated using fine-grain rules with `tolerations` field. If specifying multiple `tolerations`, pod will be scheduled if one of the tolerations is satisfied.
+
+
+#### Allowed
+
+This allows pods with tolerations the following:
```

Comment:
It's unclear from this how the matching works. Does this only match pods that have exactly this set of tolerations? Or a pod that has any of these tolerations (including none?). What if I wanted to require a toleration? I think we need to be careful how we define the semantics for things that match lists (or maps).

Reply:
From what we've said, I think that all the matches should be satisfied to return the corresponding action.

Reply:
I think I agree. It's a match if the pod has all the tolerations listed. If the pod also has other tollerations, it's still a match.

Reply:
How does this work with tolerations using set-based selectors. Does it require an exact match for each selector or does it check for equivalent selectors?

example:
For a pod:
```yaml
apiVersion: v1
kind: Pod
metadata:
  name: with-tolerations
  namespace: team
spec:
  tolerations:
  - key: "key1"
    operator: "In"
    value: ["value1a", "value1b"]
    effect: "NoSchedule"
  - key: "key2"
    operator: "In"
    value: ["value2a", "value2b"]
    effect: "NoSchedule"
  - key: "key3"
    operator: "Nin"
    value: ["value3a", "value4b"]
    effect: "NoExecute"
  containers:
  - name: with-tolerations
    image: k8s.gcr.io/pause:2.0
```

and policy:
```yaml
apiVersion: policy/v1alpha1
kind: SchedulingPolicy
metadata:
  name: mySchedulingPolicy
spec:
  action: allow   
  rules:
    namespaces:
      - key: team
        operator: Exists
    podSelector: {}    
    tolerations:
       match:
        - key: "key1"
          operator: "Nin"
          value: "value1c"
          effects: "NoSchedule"
        - key: "key2"
          operator: "Exists"
          effects: "NoSchedule"
        - key: "key2"
          operator: "Nin"
          value: ["value1a"]
          effect: "NoExecute"
```
should this policy match this pod?

Reply:
@ericavonb to me, we do not support set-based values for tolerations. But we do have this problem with nodeSelector/affinities. I think we that we can do a semantic match and not just static matching. But I’ll defer to @bsalamat and @arnaudmz

Reply:
Semantic matching could get complex for users. I think for set-based values, we can simplify and follow these rules:
1. Keys must be equal.
1. Operators must be equal.
1. `policy.spec.tolerations.match.value` must be a subset of `pod.spec.tolerations.value`

If all of the above rules match, we consider the policy a match.

@tallclair, what do you think?

Reply:
I think this is a hard problem, that deserves some thoughtful analysis and validation against real use cases.

> `policy.spec.tolerations.match.value` must be a subset of ...

This is tricky because the tolerance is sort of dependent on the operator. For example, if I want to express "allow pods that are defined to these nodes", allowing a subset makes sense. However, if I want to say "allow pods that are excluded from these nodes", I don't want to allow pods that are only excluded from a subset.

Reply:
Maybe we should just require an exact match in this case. (modulo sort)

Reply:
@tallclair - we can go down this path, subset matching can be performed by splitting into multiple policies, any thoughts @bsalamat @ericavonb  ?

Reply:
@tallclair - we can go down this path, subset matching can be performed by splitting into multiple policies, any thoughts @bsalamat @ericavonb @arnaudmz   ?